src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -702,8 +702,7 @@ void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr
     else
         tables_marked_dropped.push_back({table_id, table, dropped_metadata_path, drop_time});
     tables_marked_dropped_ids.insert(table_id.uuid);
-    /// If list of dropped tables was empty, start a drop task
-    if (drop_task && tables_marked_dropped.size() == 1)
+    if (drop_task)
```

Comment:
Could you please explain in more details, why do we need to always reschedule it here? Drop task [reschedules itself](https://github.com/ClickHouse/ClickHouse/blob/a0639d0abd996ae10033c8013a2f18ebadc1c4bf/src/Interpreters/DatabaseCatalog.cpp#L766) if there are some tables to drop in queue, and `enqueueDroppedTableCleanup(...)` reschedules the task if it just added the first table to empty queue.

Reply:
>Drop task reschedules itself if there are some tables to drop in queue

Hm, but in `DatabaseCatalog::dropTableDataTask` check the queue size only at the beginning and someone may add new entries after (while executing `dropTableFinally()`)

Reply:
New entries are added only in `enqueueDroppedTableCleanup(...)` and in `dropTableDataTask()` (when retrying failed deletion). If queue size was zero at the beginning of `dropTableDataTask()`, then task should be rescheduled when adding the first entry [here](https://github.com/ClickHouse/ClickHouse/blob/4ab39d2bee0e6f03507ba3fa1c62ba835d7a105b/src/Interpreters/DatabaseCatalog.cpp#L706) or [here](https://github.com/ClickHouse/ClickHouse/blob/4ab39d2bee0e6f03507ba3fa1c62ba835d7a105b/src/Interpreters/DatabaseCatalog.cpp#L758).

Reply:
Thanks, after looking a little bit deeper I understood how it works, will drop this patch